src/main/java/org/jenkinsci/plugins/workflow/log/FileLogStorage.java

```diff
@@ -0,0 +1,276 @@
+                    long r = this.writeRawLogTo(start, new FilterOutputStream(caos) {
+                        long lastTransition = -1;
+                        boolean eof;
+                        String lastId;
+                        long pos = start;
+                        boolean hadLastId;
+                        @Override public void write(int b) throws IOException {
+                            while (lastTransition < pos && !eof) {
+                                String line = indexBR.readLine();
+                                if (line == null) {
+                                    eof = true;
+                                    break;
+                                }
+                                int space = line.indexOf(' ');
+                                try {
+                                    lastTransition = Long.parseLong(space == -1 ? line : line.substring(0, space));
+                                } catch (NumberFormatException x) {
+                                    LOGGER.warning("Ignoring corrupt index file " + index);
+                                }
+                                lastId = space == -1 ? null : line.substring(space + 1);
+                            }
+                            if (pos == lastTransition) {
+                                if (hadLastId) {
+                                    w.write(LogStorage.endStep());
+                                }
+                                hadLastId = lastId != null;
+                                if (lastId != null) {
+                                    w.write(LogStorage.startStep(lastId));
+                                }
+                            }
+                            super.write(b);
+                            pos++;
+                        }
+                        @Override public void flush() throws IOException {
+                            if (lastId != null) {
+                                w.write(LogStorage.endStep());
+                            }
+                            super.flush();
+                        }
+                    });
+                    ConsoleAnnotators.setAnnotator(caos.getConsoleAnnotator());
+                    return r;
+                }
+            }
+        };
+    }
+
+    @Override public AnnotatedLargeText<FlowNode> stepLog(FlowNode node, boolean complete) {
+        String id = node.getId();
+        try (ByteBuffer buf = new ByteBuffer();
+             RandomAccessFile raf = new RandomAccessFile(log, "r");
+             BufferedReader indexBR = index.isFile() ? Files.newBufferedReader(index.toPath(), StandardCharsets.UTF_8) : new BufferedReader(new NullReader(0))) {
+            String line;
+            long pos = -1; // -1 if not currently in this node, start position if we are
+            while ((line = indexBR.readLine()) != null) {
```

Comment:
For those keeping track at home: any time we try to read a single node's logs we trigger a full read of the entire index file, and as we scan we look for lines with our FlowNode ID.  Which thankfully is *smaller* than the *entire* logfile by probably orders of magnitude. 

But still this feels both overly complicated and inefficient - with `O(m)` performance to find all a FlowNode's logs, where `m` is the size of the index.   Is there a reason we can't simply store when FlowNode is "done" for logging purposes -- we have that logic handy already.

Reply:
> store when FlowNode is "done"

Not sure what you are proposing exactly.

Reply:
@jglick We know how to compute when a FlowNode is "closed" for logging, depending on if it's an AtomNode, StartNode, or EndNode.   This limits the scope we'd need to scan. 

Now, if we specially marked the offset for the first and last log content for a FlowNode, and stored it specially (say as fields in our LogAction implementation, or showing boundaries a bit more clearly in the index) we could further limit the amount of scanning done. 

However... eh, this can be optimized later.

Reply:
> stored it specially (say as fields in our `LogAction` implementation

I preferred to avoid duplicating information like this. In other words, `log` + `log-index` should be self-contained.

Now another line type (not a transition) _could_ be added to `log-index` indicating that a given `FlowNode` has now been closed and that thus there is no need to search past that point in the file looking for subsequent blocks. The extra complexity just to avoid reading the rest of what would usually be a short text file (smaller than a typical “disk” block size I think) does not seem worth it.

Reply:
Please leave this comment *unresolved* so we know in the future where to start if we need to optimize.